### DISCUSSION TITLE
Can't I use `pyeval` with multiple statements?

### DISCUSSION THREAD

**Discussion by jkrumbiegel** [cjdoris/PythonCall.jl#551] [General]
I was looking into supporting python cells in QuartoNotebookRunner using PythonCall. A python cell in the context of a Julia notebook would mean that PythonCall would handle executing the python code in the cell and then return the last object converted to Julia.

I've found `pyeval` but it seems not to support multiple statements:

```julia
julia> module PyModule end
julia> pyeval(Any, """
       x = 1 + 1
       1 + 1
       """, PyModule)
ERROR: Python: SyntaxError: invalid syntax (<string>, line 1)
Python stacktrace: none
Stacktrace:
 [1] pythrow()
```

I can run such code blocks with `pyexec` but this doesn't return the last result. One hacky workaround would be to split the code I get into parts where only the last statement is `pyeval`ed, but then I'd have to analyze the code which seems very error prone. So I'm hoping there's maybe a different way to use `pyeval` that I haven't yet found?

**Comment by cjdoris** [comment]
Correct, `pyeval` is for evaluating a single expression, `pyexec` is for executing statements. In Python, only expressions have values (unlike Julia where everything is an expression that takes a value).

So you need to use `pyexec` or `@pyexec`, assign the value you want to extract to a variable in the code, and there are mechanisms to extract the variable in the call to `pyexec`/`@pyexec` (see the docstrings). For example 
```julia
@pyexec """
x = 1+1
y = 1+1
""" => y
```

**Reply by jkrumbiegel** [reply]
The problem for me is that I'm writing a generic mechanism which people use with their own python code. And it's supposed to act like an ipython notebook cell where I think you'll also sometimes get a value back and sometimes not, depending on the type of expression? So I don't know what code I'm going to get and which pyeval or pyexec might be appropriate.

**Reply by cjdoris** [reply]
You could use the `ast` stdlib to parse the Python code and see if the last statement is an expression. Ultimately though none of this is much to do with PythonCall - you need to figure out how to do it in Python first then translate it to PythonCall.

**Reply by jkrumbiegel** [reply]
Hm ok seems it's more complicated than I thought. Thank you for your feedback, I might try the ast approach at some point

**Comment by asinghvi17** [comment]
From the docstring and some basic tests, it looks like running code in `pyexec` does not save variables to either global or local scope.  Is there an option where we could enable that without modifying the Python code?  

If running an interactive session, for example, I would expect that variables are propagated from one cell to the next, without having to append `global` to every invocation.

**Reply by cjdoris** [reply]
The behaviour of `pyexec` and `pyeval` is basically identical to their python counterparts `exec` and `eval` which have the same issues.

That is, if you pass both a dict of globals and a dict of locals, any variables set are put in the locals. If you just pass globals, variables are put there.

**Reply by asinghvi17** [reply]
Aha, thanks!  Will try that out - if it works life will be substantially easier :)
